File: typing_trainer.py

#!/usr/bin/env python3
# typing_trainer.py
import argparse
import random
import time
import sys
import json
import csv
from colorama import init, Fore, Style

init(autoreset=True)

class TypingTrainer:
    def __init__(self, length=50, charset='all', mode='char', language='en'):
        self.length = length
        self.mode = mode
        self.language = language
        self.charset = self._get_charset(charset)
        self.text = self._generate_text()
        self.errors = 0
        self.total_chars = 0
        self.start_time = None
        self.end_time = None

    def _get_charset(self, charset):
        if self.language == 'ru':
            letters = 'абвгдеёжзийклмнопрстуфхцчшщъыьэюя'
            letters += letters.upper()
        else:
            letters = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ'
        digits = '0123456789'
        symbols = '!@#$%^&*()_+-=,.;:?/[]{}|'
        if charset == 'letters':
            return letters
        elif charset == 'alnum':
            return letters + digits
        else:
            return letters + digits + symbols

    def _generate_text(self):
        if self.mode == 'word':
            words = []
            for _ in range(max(1, self.length // 5)):
                word_len = random.randint(3, 8)
                word = ''.join(random.choice(self.charset) for _ in range(word_len))
                words.append(word)
            return ' '.join(words)
        elif self.mode == 'sentence':
            sentences = []
            for _ in range(max(1, self.length // 30)):
                words_count = random.randint(5, 12)
                words = []
                for _ in range(words_count):
                    wl = random.randint(3, 8)
                    words.append(''.join(random.choice(self.charset) for _ in range(wl)))
                sentence = ' '.join(words).capitalize() + '.'
                sentences.append(sentence)
            return ' '.join(sentences)
        else:
            return ''.join(random.choice(self.charset) for _ in range(self.length))

    def run(self):
        print(Fore.CYAN + "Тренажер печати (символы)")
        print(Fore.YELLOW + "Введите следующий текст как можно точнее и быстрее:")
        print(Fore.WHITE + self.text)
        print("\nНажмите Enter, чтобы начать...")
        input()
        self.start_time = time.time()
        user_input = input(Fore.GREEN + "> ")
        self.end_time = time.time()
        self.total_chars = len(user_input)
        # Сравнение
        min_len = min(len(self.text), len(user_input))
        self.errors = sum(1 for i in range(min_len) if self.text[i] != user_input[i])
        if len(user_input) > len(self.text):
            self.errors += len(user_input) - len(self.text)
        # Вывод результатов
        elapsed = self.end_time - self.start_time
        cpm = (self.total_chars / elapsed) * 60 if elapsed > 0 else 0
        wpm = (self.total_chars / 5 / elapsed) * 60 if elapsed > 0 else 0
        accuracy = ((self.total_chars - self.errors) / self.total_chars * 100) if self.total_chars > 0 else 0
        print(Fore.CYAN + "\n--- Результаты ---")
        print(Fore.GREEN + f"Скорость: {cpm:.1f} симв/мин ({wpm:.1f} слов/мин)")
        print(Fore.YELLOW + f"Точность: {accuracy:.1f}%")
        print(Fore.MAGENTA + f"Время: {elapsed:.2f} сек")
        print(Fore.RED + f"Ошибок: {self.errors}")
        return {"cpm": cpm, "wpm": wpm, "accuracy": accuracy, "time": elapsed, "errors": self.errors}

def main():
    parser = argparse.ArgumentParser(description="Тренажер печати")
    parser.add_argument("--length", type=int, default=50, help="Длина текста")
    parser.add_argument("--charset", choices=["letters", "alnum", "all"], default="all")
    parser.add_argument("--mode", choices=["char", "word", "sentence"], default="char")
    parser.add_argument("--language", choices=["en", "ru"], default="en")
    parser.add_argument("--export-stats", help="Экспорт статистики в файл (JSON или CSV)")
    args = parser.parse_args()

    trainer = TypingTrainer(args.length, args.charset, args.mode, args.language)
    stats = trainer.run()
    if args.export_stats:
        ext = args.export_stats.split('.')[-1].lower()
        if ext == 'json':
            with open(args.export_stats, 'w') as f:
                json.dump(stats, f, indent=2)
        elif ext == 'csv':
            with open(args.export_stats, 'w', newline='') as f:
                writer = csv.DictWriter(f, fieldnames=stats.keys())
                writer.writeheader()
                writer.writerow(stats)
        print(f"Статистика сохранена в {args.export_stats}")

if __name__ == "__main__":
    main()

File: typing_trainer.js
#!/usr/bin/env node
// typing_trainer.js
const { program } = require('commander');
const readline = require('readline');
const chalk = require('chalk');
const fs = require('fs');

class TypingTrainer {
    constructor(length = 50, charset = 'all', mode = 'char', language = 'en') {
        this.length = length;
        this.mode = mode;
        this.language = language;
        this.charset = this._getCharset(charset);
        this.text = this._generateText();
        this.startTime = null;
        this.endTime = null;
        this.errors = 0;
        this.totalChars = 0;
        this.rl = readline.createInterface({ input: process.stdin, output: process.stdout });
    }

    _getCharset(charset) {
        let letters = this.language === 'ru' ?
            'абвгдеёжзийклмнопрстуфхцчшщъыьэюяАБВГДЕЁЖЗИЙКЛМНОПРСТУФХЦЧШЩЪЫЬЭЮЯ' :
            'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ';
        const digits = '0123456789';
        const symbols = '!@#$%^&*()_+-=,.;:?/[]{}|';
        if (charset === 'letters') return letters;
        if (charset === 'alnum') return letters + digits;
        return letters + digits + symbols;
    }

    _generateText() {
        if (this.mode === 'word') {
            const words = [];
            const count = Math.max(1, Math.floor(this.length / 5));
            for (let i = 0; i < count; i++) {
                const wl = 3 + Math.floor(Math.random() * 6);
                let word = '';
                for (let j = 0; j < wl; j++) word += this.charset[Math.floor(Math.random() * this.charset.length)];
                words.push(word);
            }
            return words.join(' ');
        } else if (this.mode === 'sentence') {
            const sentences = [];
            const count = Math.max(1, Math.floor(this.length / 30));
            for (let i = 0; i < count; i++) {
                const wc = 5 + Math.floor(Math.random() * 8);
                const words = [];
                for (let j = 0; j < wc; j++) {
                    const wl = 3 + Math.floor(Math.random() * 6);
                    let word = '';
                    for (let k = 0; k < wl; k++) word += this.charset[Math.floor(Math.random() * this.charset.length)];
                    words.push(word);
                }
                let sentence = words.join(' ');
                sentence = sentence.charAt(0).toUpperCase() + sentence.slice(1) + '.';
                sentences.push(sentence);
            }
            return sentences.join(' ');
        } else {
            let text = '';
            for (let i = 0; i < this.length; i++) {
                text += this.charset[Math.floor(Math.random() * this.charset.length)];
            }
            return text;
        }
    }

    async run() {
        console.log(chalk.cyan('Тренажер печати (символы)'));
        console.log(chalk.yellow('Введите следующий текст как можно точнее и быстрее:'));
        console.log(chalk.white(this.text));
        console.log('\nНажмите Enter, чтобы начать...');
        await this._question('');
        this.startTime = Date.now();
        const userInput = await this._question(chalk.green('> '));
        this.endTime = Date.now();
        this.totalChars = userInput.length;
        const minLen = Math.min(this.text.length, userInput.length);
        this.errors = 0;
        for (let i = 0; i < minLen; i++) {
            if (this.text[i] !== userInput[i]) this.errors++;
        }
        if (userInput.length > this.text.length) this.errors += userInput.length - this.text.length;
        const elapsed = (this.endTime - this.startTime) / 1000;
        const cpm = (this.totalChars / elapsed) * 60;
        const wpm = (this.totalChars / 5 / elapsed) * 60;
        const accuracy = this.totalChars > 0 ? ((this.totalChars - this.errors) / this.totalChars * 100) : 0;
        console.log(chalk.cyan('\n--- Результаты ---'));
        console.log(chalk.green(`Скорость: ${cpm.toFixed(1)} симв/мин (${wpm.toFixed(1)} слов/мин)`));
        console.log(chalk.yellow(`Точность: ${accuracy.toFixed(1)}%`));
        console.log(chalk.magenta(`Время: ${elapsed.toFixed(2)} сек`));
        console.log(chalk.red(`Ошибок: ${this.errors}`));
        this.rl.close();
        return { cpm, wpm, accuracy, time: elapsed, errors: this.errors };
    }

    _question(prompt) {
        return new Promise(resolve => this.rl.question(prompt, resolve));
    }
}

program
    .option('-l, --length <n>', 'Длина текста', parseInt, 50)
    .option('-c, --charset <set>', 'letters, alnum, all', 'all')
    .option('-m, --mode <mode>', 'char, word, sentence', 'char')
    .option('-L, --language <lang>', 'en, ru', 'en')
    .option('--export-stats <file>', 'Экспорт статистики')
    .parse(process.argv);

const opts = program.opts();
const trainer = new TypingTrainer(opts.length, opts.charset, opts.mode, opts.language);
trainer.run().then(stats => {
    if (opts.exportStats) {
        const ext = opts.exportStats.split('.').pop().toLowerCase();
        if (ext === 'json') fs.writeFileSync(opts.exportStats, JSON.stringify(stats, null, 2));
        else if (ext === 'csv') {
            const header = Object.keys(stats).join(',') + '\n';
            const row = Object.values(stats).join(',') + '\n';
            fs.writeFileSync(opts.exportStats, header + row);
        }
        console.log(`Статистика сохранена в ${opts.exportStats}`);
    }
});
